stl/inc/ranges

```diff
@@ -636,12 +636,138 @@ namespace ranges {
         /* [[no_unique_address]] */ _Ty _Value{};
     };
 
-    // CLASS TEMPLATE ranges::empty_view
+    // CLASS TEMPLATE ranges::_Non_propagating_cache
     // clang-format off
-    template <class _Ty>
-        requires is_object_v<_Ty>
-    class empty_view : public view_interface<empty_view<_Ty>> {
+    template <_Destructible_object _Ty>
+    class _Non_propagating_cache { // a simplified optional that resets on copy / move
+        // clang-format on
+    public:
+        constexpr _Non_propagating_cache() noexcept : _Dummy{} {}
+
+        _CONSTEXPR20 ~_Non_propagating_cache() noexcept {
+            if (_Engaged) {
+                _Val.~_Ty();
+                _Construct_in_place(_Dummy);
+            }
+        }
+
+        // clang-format off
+        ~_Non_propagating_cache() requires is_trivially_destructible_v<_Ty> = default;
+        // clang-format on
+
+        constexpr _Non_propagating_cache(const _Non_propagating_cache&) noexcept : _Dummy{} {}
+
+        _CONSTEXPR20 _Non_propagating_cache(_Non_propagating_cache&& _Other) noexcept : _Dummy{} {
+            if (_Other._Engaged) {
+                _Other._Val.~_Ty();
+                _Construct_in_place(_Other._Dummy);
+                _Other._Engaged = false;
+            }
+        }
+
+        _CONSTEXPR20 _Non_propagating_cache& operator=(const _Non_propagating_cache& _Other) noexcept {
+            if (_STD addressof(_Other) == this) {
+                return *this;
+            }
+
+            if (_Engaged) {
+                _Val.~_Ty();
+                _Construct_in_place(_Dummy);
+                _Engaged = false;
+            }
+
+            return *this;
+        }
+
+        _CONSTEXPR20 _Non_propagating_cache& operator=(_Non_propagating_cache&& _Other) noexcept {
+            if (_Engaged) {
+                _Val.~_Ty();
+                _Construct_in_place(_Dummy);
+                _Engaged = false;
+            }
+
+            if (_Other._Engaged) {
+                _Other._Val.~_Ty();
+                _Construct_in_place(_Other._Dummy);
+                _Other._Engaged = false;
+            }
+
+            return *this;
+        }
+
+        _NODISCARD constexpr _Ty& operator*() noexcept {
+            _STL_INTERNAL_CHECK(_Engaged);
+            return _Val;
+        }
+        _NODISCARD constexpr const _Ty& operator*() const noexcept {
+            _STL_INTERNAL_CHECK(_Engaged);
+            return _Val;
+        }
+
+        template <class _It>
+        _CONSTEXPR20 _Ty& _Emplace_deref(_It& _Iter) noexcept(noexcept(_Ty(*_Iter))) /* strengthened */ {

```

Comment:
`_Ty(*_Iter)` is a functional-style C-semantics cast. Can we rephrase this in terms of `_Construct_in_place(_Val, *_Iter)` which has conditional `noexcept`, or directly in terms of `is_nothrow_constructible_v`?